### PR TITLE
Store Overhaul

### PR DESCRIPTION
**Pull request by ghost**
- [x] Update Store To Sell Tokens.

- [x] Update Slots To Use Tokens

Before Merge Add a Entry Named tokenPrice to conf

### CODE REVIEW

File: commands/social/games/slots.js [hunk 1]
```diff
@@ -33,9 +33,11 @@ class Slots extends Social {
   async run(message, args, level) { // eslint-disable-line no-unused-vars
     if (message.settings.socialSystem !== "true") return message.response(undefined, "The social system has been disabled.");
   
-    if (!(await this.cmdPay(message, message.author.id, this.help.cost))) return;
+    if (!message.member.inventory["tokens"]) return message.response(undefined, "Ba...Baka You don't have any tokens to play Slots, buy some with the Store command");
```
Comment (MrJacz) [inline commands/social/games/slots.js]: use `.tokens`

File: commands/social/store/store.js [hunk 2]
```diff
@@ -1,12 +1,98 @@
 const Social = require(`${process.cwd()}/base/Social.js`);
 
+const buyToken = async (name, message, client) => {
```
Comment (MrJacz) [inline commands/social/store/store.js]: Make this a function inside the class...

File: commands/social/store/store.js [hunk 3]
```diff
@@ -1,12 +1,98 @@
...
+    message.channel.send("You have bought a token");
+  
+  } else
+  
+  if (["n", "no", "cancel"].includes(response.toLowerCase())) {
+    message.response(undefined, "Transaction cancelled.");
+  }
+};
+
+const buyRole = async (name, message, client) => {
```
Comment (MrJacz) [inline commands/social/store/store.js]: same here

File: commands/social/store/store.js [hunk 4]
```diff
@@ -1,12 +1,98 @@
...
+    message.channel.send("You have bought the role :tada: ");
+  
+  } else
+  
+  if (["n", "no", "cancel"].includes(response.toLowerCase())) {
+    message.response(undefined, "Transaction cancelled.");
+  }
+};
+
+const sellToken = async (name, message, client) => {
```
Comment (MrJacz) [inline commands/social/store/store.js]: same here

File: commands/social/store/store.js [hunk 5]
```diff
@@ -1,12 +1,98 @@
...
+    message.channel.send("You have sold a token");
+  
+  } else
+  
+  if (["n", "no", "cancel"].includes(response.toLowerCase())) {
+    message.response(undefined, "Transaction cancelled.");
+  }
+};
+
+const sellRole = async (name, message, client) => {
```
Comment (MrJacz) [inline commands/social/store/store.js]: same here

File: commands/social/store/store.js [hunk 6]
```diff
@@ -21,59 +107,22 @@ class Store extends Social {
     switch (message.flags[0]) {
       case ("buy"): {
```
Comment (MrJacz) [inline commands/social/store/store.js]: the params around `"buy"` arent needed

File: commands/social/store/store.js [hunk 7]
```diff
@@ -21,59 +107,22 @@ class Store extends Social {
...
+        if (!name) return message.reply(`|\`❌\`| ${this.help.usage}`);
+        if (name.toLowerCase() === "slots") {
+          buyToken(name, message, this.client);
+        } else {
+          buyRole(name, message, this.client);
         }
         break;
       }
 
       case ("sell"): {
```
Comment (MrJacz) [inline commands/social/store/store.js]: same here

File: commands/social/store/store.js [hunk 8]
```diff
@@ -106,23 +155,23 @@ class Store extends Social {
...
         }
         break;
       }
 
       case ("view"): {
         const items = message.guild.store;
-        if (items.length === 0) return message.response(undefined, "Baka... nothing is for sale!");
-        message.channel.send(`= ${message.guild.name} General Store =\n` + items.map(item => 
-          `${message.guild.roles.get(item.id.toString()).name}${" ".repeat(40 - message.guild.roles.get(item.id.toString()).name.length)}::  ${parseInt(item.price) === 0 ? "FREE" : `₲${parseInt(item.price).toLocaleString()}`} :: ${message.member.roles.has(item.id) ? "✓" : ""}`).join("\n"), { code: "asciidoc" }
+        if (items.length === 0) return  message.channel.send(`= ${message.guild.name} Store =\n` + "\n" + "= Tokens = \n" + `Slots${" ".repeat(20 - "Slots".length)}:: ₲${message.settings.tokenPrice} :: ${message.member.inventory["tokens"]} \n`, { code: "asciidoc" });
```
Comment (MrJacz) [inline commands/social/store/store.js]: `if (!items.length)`

**Review by MrJacz — CHANGES_REQUESTED** [review]
Just a few

File: commands/social/store/store.js (hunk 6 above)
Comment (ghost) [reply commands/social/store/store.js]: params around buy ?

🤔

Comment (MrJacz) [reply commands/social/store/store.js]: `()`

**Comment by YorkAARGH** [comment]
Make sure you resolve merge conflicts.

**Comment by ghost** [comment]
Fixed Merge Conflicts ✔️

File: commands/social/games/slots.js [hunk 9]
```diff
@@ -32,16 +32,22 @@ class Slots extends Social {
...
+      const results = machine.play();
+      const winnings = this.help.cost * results.totalPoints;
+      message.buildEmbed()
+        .setColor(message.guild.me.roles.highest.color || 5198940)
+        .setAuthor("Misaki Slots")
+        .setDescription(`${results.visualize(false)}\n\n${results.winCount === 0 ? `${message.member.displayName} has lost!\nBetter luck next time!` : `Whoa... ${message.member.displayName} won!`}\n\n${results.winCount === 0 ? "" : `You have won ₲${winnings.toLocaleString()}`}`)
+        .setTimestamp()
+        .send();
+      if (results.winCount > 0) return message.member.givePoints(winnings);
+    } catch (e) {
```
Comment (DevYukine) [inline commands/social/games/slots.js]: i removed all try/catch in there for a duplication reason please do not change that

File: commands/social/store/store.js [hunk 10]
```diff
@@ -7,6 +7,7 @@ class Store extends Social {
       description: "Display All Store Items",
       usage: "store <-buy|-sell|-add|-del|-view>",
       category: "Social",
+      aliases: []
```
Comment (DevYukine) [inline commands/social/store/store.js]: this is unnecessary

File: commands/social/store/store.js [hunk 11]
```diff
@@ -106,26 +70,109 @@ class Store extends Social {
...
+        if (!items.length) return  message.channel.send(`= ${message.guild.name} Store =\n` + "\n" + "= Tokens = \n" + `Slots${" ".repeat(20 - "Slots".length)}:: ₲${message.settings.tokenPrice} :: ${message.member.inventory["tokens"]} \n`, { code: "asciidoc" });
+        message.channel.send(`= ${message.guild.name} Store =\n` + "\n" + "= Tokens = \n" + `Slots${" ".repeat(20 - "Slots".length)}:: ₲${message.settings.tokenPrice} :: ${message.member.inventory["tokens"]} \n \n`+ "= Roles For Sale = \n"+ items.map(item => 
+          `${message.guild.roles.get(item.id.toString()).name}${" ".repeat(20 - message.guild.roles.get(item.id.toString()).name.length)}::  ${parseInt(item.price) === 0 ? "FREE" : `₲${parseInt(item.price).toLocaleString()}`} ${message.member.roles.has(item.id) ? ":: ✓" : ""}`).join("\n"), { code: "asciidoc" }
         );
       }
     }
   }
+  async buyToken(name, message, client) {
+    if (parseInt(message.settings.tokenPrice) > parseInt(message.member.score.points)) {
```
Comment (DevYukine) [inline commands/social/store/store.js]: i would personal recommend using the Number constructor over parseInt

File: commands/social/store/store.js [hunk 12]
```diff
@@ -106,26 +70,109 @@ class Store extends Social {
...
+      return message.channel.send(`You currently have ₲${parseInt(message.member.score.points).toLocaleString()}, but the token costs ${parseInt(message.settings.tokenPrice).toLocaleString()}!`);
```
Comment (DevYukine) [inline commands/social/store/store.js]: same as in line 95

File: commands/social/store/store.js [hunk 13]
```diff
@@ -106,26 +70,109 @@ class Store extends Social {
...
+    }
+  }
+  async buyRole(name, message, client) {
+    const item = client.store.filter(i => i.name.toLowerCase().includes(name.toLowerCase()));
+    
+    if (item.size > 1) return message.reply(`B..Baka! Be more specific more than that, there is more than one item on sale with ${name} as their name`);
+    if (!item) return message.channel.send("That item doesn't exist, Please make sure it is spelled correctly");
+    if (message.member.roles.has(item.array()[0].id)) return message.channel.send("You already have the role :facepalm: ");
+    
+    if (parseInt(item.array()[0].price) > parseInt(message.member.score.points)) {
```
Comment (DevYukine) [inline commands/social/store/store.js]: transform that collection to an array is pointless here and uses more time and resources aswell as what i mentioned in line 95 again

File: commands/social/store/store.js [hunk 14]
```diff
@@ -106,26 +70,109 @@ class Store extends Social {
...
+      return message.channel.send(`You currently have ₲${parseInt(message.member.score.points).toLocaleString()}, but the role costs ${parseInt(item.array()[0].price).toLocaleString()}!`);
+    }
+  
+    const filter = m => m.author.id === message.author.id;
+    const response = await client.awaitReply(message, `Are you sure you want to purchase ${item.array()[0].name} for ₲${item.array()[0].price.toLocaleString()}?`, filter, undefined, null);
+    if (["y", "yes"].includes(response.toLowerCase())) {
+    
+      message.member.takePoints(parseInt(item.array()[0].price));
```
Comment (DevYukine) [inline commands/social/store/store.js]: same as in line 119

File: commands/social/store/store.js [hunk 15]
```diff
@@ -106,26 +70,109 @@ class Store extends Social {
...
+      await message.member.roles.add(item.array()[0].id);
```
Comment (DevYukine) [inline commands/social/store/store.js]: same as in line 119

File: commands/social/store/store.js [hunk 16]
```diff
@@ -106,26 +70,109 @@ class Store extends Social {
...
+      message.channel.send("You have bought the role :tada: ");
+    
+    } else
+    
+    if (["n", "no", "cancel"].includes(response.toLowerCase())) {
+      message.response(undefined, "Transaction cancelled.");
+    }
+  }
+  async sellToken(name, message, client) {
+    const returnPrice = parseInt(message.settings.tokenPrice) / 2;
```
Comment (DevYukine) [inline commands/social/store/store.js]: same as in line 95

File: commands/social/store/store.js [hunk 17]
```diff
@@ -106,26 +70,109 @@ class Store extends Social {
...
+    if (["n", "no", "cancel"].includes(response.toLowerCase())) {
+      message.response(undefined, "Transaction cancelled.");
+    }
+  }
+  
+  async sellRole(name, message, client) {
+    const item = client.store.filter(i => i.name.toLowerCase().includes(name.toLowerCase()));
+          
+    if (!item) return message.channel.send("That item doesn't exist, Please make sure it is spelled correctly");
+    if (!message.member.roles.has(item.array()[0].id)) return message.channel.send("You don't have the role :facepalm: ");
```
Comment (DevYukine) [inline commands/social/store/store.js]: same as in line 119

File: commands/social/store/store.js [hunk 18]
```diff
@@ -106,26 +70,109 @@ class Store extends Social {
...
+    
+    const returnPrice = Math.floor(item.array()[0].price/2);
```
Comment (DevYukine) [inline commands/social/store/store.js]: same as in line 119

File: commands/social/store/store.js [hunk 19]
```diff
@@ -106,26 +70,109 @@ class Store extends Social {
...
+    
+    const filter = m => m.author.id === message.author.id;
+    const response = await client.awaitReply(message, `Are you sure you want to sell ${item.array()[0].name} for ₲${returnPrice.toLocaleString()}?`, filter, undefined, null);
+    if (["y", "yes"].includes(response.toLowerCase())) {
+    
+      message.member.givePoints(returnPrice);
+      await message.member.roles.remove(item.array()[0].id);
```
Comment (DevYukine) [inline commands/social/store/store.js]: same as in line 119

File: commands/social/store/store.js [hunk 20]
```diff
@@ -19,70 +20,33 @@ class Store extends Social {
...
-          await message.member.roles.add(item.array()[0].id);
-          message.channel.send("You have bought the role :tada: ");
-        
-        } else
-        
-        if (["n", "no", "cancel"].includes(response.toLowerCase())) {
-          message.response(undefined, "Transaction cancelled.");
+        if (!name) return message.reply(`|\`❌\`| ${this.help.usage}`);
+        if (name.toLowerCase() === "slots") {
+          this.buyToken(name, message, this.client);
```
Comment (DevYukine) [inline commands/social/store/store.js]: please await all methods which return a Promise in here this should be done in the whole PR

**Review by DevYukine — CHANGES_REQUESTED** [review]

**Comment by ghost** [comment]
@Dev-Yukine I have tried and fixed everything you asked for :)

File: commands/social/store/store.js [hunk 21]
```diff
@@ -106,26 +69,109 @@ class Store extends Social {
...
         } else
         
-        if (["n","no","cancel"].includes(response.toLowerCase())) {
+        if (["n","no","cancel"].includes(response)) {
           message.reply("Action cancelled.");
         }
         break;
       }
 
       case ("view"): {
```
Comment (MrJacz) [inline commands/social/store/store.js]: Literally no need to put the parentheses there

File: commands/social/store/store.js [hunk 22]
```diff
@@ -106,26 +69,109 @@ class Store extends Social {
...
+        if (!items.length) return  message.channel.send(`= ${message.guild.name} Store =\n` + "\n" + "= Tokens = \n" + `Slots${" ".repeat(20 - "Slots".length)}:: ₲${message.settings.tokenPrice} :: ${message.member.inventory["tokens"]} \n`, { code: "asciidoc" });
+        message.channel.send(`= ${message.guild.name} Store =\n` + "\n" + "= Tokens = \n" + `Slots${" ".repeat(20 - "Slots".length)}:: ₲${message.settings.tokenPrice} :: ${message.member.inventory["tokens"]} \n \n`+ "= Roles For Sale = \n"+ items.map(item => 
+          `${message.guild.roles.get(item.id.toString()).name}${" ".repeat(20 - message.guild.roles.get(item.id.toString()).name.length)}::  ${Number(item.price) === 0 ? "FREE" : `₲${Number(item.price).toLocaleString()}`} ${message.member.roles.has(item.id) ? ":: ✓" : ""}`).join("\n"), { code: "asciidoc" }
         );
       }
     }
   }
+  async buyToken(name, message, client) {
```
Comment (MrJacz) [inline commands/social/store/store.js]: you shouldn't be passing client, you can still access it with `this.client`... you're in the class

File: commands/social/store/store.js [hunk 23]
```diff
@@ -106,26 +69,109 @@ class Store extends Social {
...
+      await message.member.giveItem("tokens", 1);
+      message.channel.send("You have bought a token");
+    
+    } else
+    
+    if (["n", "no", "cancel"].includes(response.toLowerCase())) {
+      message.response(undefined, "Transaction cancelled.");
+    }
+  }
+  async buyRole(name, message, client) {
```
Comment (MrJacz) [inline commands/social/store/store.js]: you shouldn't be passing client, you can still access it with `this.client`... you're in the class

File: commands/social/store/store.js [hunk 24]
```diff
@@ -106,26 +69,109 @@ class Store extends Social {
...
+      await message.member.roles.add(item.first().id);
+      message.channel.send("You have bought the role :tada: ");
+    
+    } else
+    
+    if (["n", "no", "cancel"].includes(response.toLowerCase())) {
+      message.response(undefined, "Transaction cancelled.");
+    }
+  }
+  async sellToken(name, message, client) {
```
Comment (MrJacz) [inline commands/social/store/store.js]: you shouldn't be passing client, you can still access it with `this.client`... you're in the class

File: commands/social/store/store.js [hunk 25]
```diff
@@ -106,26 +69,109 @@ class Store extends Social {
...
+      message.channel.send("You have sold a token");
+    
+    } else
+    
+    if (["n", "no", "cancel"].includes(response.toLowerCase())) {
+      message.response(undefined, "Transaction cancelled.");
+    }
+  }
+  
+  async sellRole(name, message, client) {
```
Comment (MrJacz) [inline commands/social/store/store.js]: you shouldn't be passing client, you can still access it with `this.client`... you're in the class

**Review by MrJacz — CHANGES_REQUESTED** [review]

File: commands/social/store/store.js [hunk 26]
```diff
@@ -106,26 +69,109 @@ class Store extends Social {
...
         }
         break;
       }
 
       case ("view"): {
         const items = message.guild.store;
-        if (items.length === 0) return message.response(undefined, "Baka... nothing is for sale!");
-        message.channel.send(`= ${message.guild.name} General Store =\n` + items.map(item => 
-          `${message.guild.roles.get(item.id.toString()).name}${" ".repeat(40 - message.guild.roles.get(item.id.toString()).name.length)}::  ${parseInt(item.price) === 0 ? "FREE" : `₲${parseInt(item.price).toLocaleString()}`} :: ${message.member.roles.has(item.id) ? "✓" : ""}`).join("\n"), { code: "asciidoc" }
+        if (!items.length) return  message.channel.send(`= ${message.guild.name} Store =\n` + "\n" + "= Tokens = \n" + `Slots${" ".repeat(20 - "Slots".length)}:: ₲${message.settings.tokenPrice} :: ${message.member.inventory["tokens"]} \n`, { code: "asciidoc" });
```
Comment (MrJacz) [inline commands/social/store/store.js]: `message.member.inventory["tokens"]` to `message.member.inventory.tokens`

File: commands/social/store/store.js [hunk 27]
```diff
@@ -106,26 +69,109 @@ class Store extends Social {
...
+        message.channel.send(`= ${message.guild.name} Store =\n` + "\n" + "= Tokens = \n" + `Slots${" ".repeat(20 - "Slots".length)}:: ₲${message.settings.tokenPrice} :: ${message.member.inventory["tokens"]} \n \n`+ "= Roles For Sale = \n"+ items.map(item =>
```
Comment (MrJacz) [inline commands/social/store/store.js]: `message.member.inventory["tokens"]` to `message.member.inventory.tokens`

**Review by MrJacz — CHANGES_REQUESTED** [review]

File: commands/social/store/store.js [hunk 28]
```diff
@@ -106,26 +69,109 @@ class Store extends Social {
...
+  }
+  async buyRole(name, message) {
+    const item = this.client.store.filter(i => i.name.toLowerCase().includes(name.toLowerCase()));
+    
+    if (item.size > 1) return message.reply(`B..Baka! Be more specific more than that, there is more than one item on sale with ${name} as their name`);
+    if (!item) return message.channel.send("That item doesn't exist, Please make sure it is spelled correctly");
+    if (message.member.roles.has(item.first().id)) return message.channel.send("You already have the role :facepalm: ");
+    
+    if (Number(item.first().price) > Number(message.member.score.points)) {
+      return message.channel.send(`You currently have ₲${Number(message.member.score.points).toLocaleString()}, but the role costs ${Number(item.first().price).toLocaleString()}!`);
```
Comment (DevYukine) [inline commands/social/store/store.js]: same as in line 94.

File: commands/social/store/store.js [hunk 29]
```diff
@@ -106,26 +69,109 @@ class Store extends Social {
...
+        if (!items.length) return  message.channel.send(`= ${message.guild.name} Store =\n` + "\n" + "= Tokens = \n" + `Slots${" ".repeat(20 - "Slots".length)}:: ₲${message.settings.tokenPrice} :: ${message.member.inventory.tokens} \n`, { code: "asciidoc" });
+        message.channel.send(`= ${message.guild.name} Store =\n` + "\n" + "= Tokens = \n" + `Slots${" ".repeat(20 - "Slots".length)}:: ₲${message.settings.tokenPrice} :: ${message.member.inventory.tokens} \n \n`+ "= Roles For Sale = \n"+ items.map(item => 
+          `${message.guild.roles.get(item.id.toString()).name}${" ".repeat(20 - message.guild.roles.get(item.id.toString()).name.length)}::  ${Number(item.price) === 0 ? "FREE" : `₲${Number(item.price).toLocaleString()}`} ${message.member.roles.has(item.id) ? ":: ✓" : ""}`).join("\n"), { code: "asciidoc" }
         );
       }
     }
   }
+  async buyToken(name, message) {
+    if (Number(message.settings.tokenPrice) > Number(message.member.score.points)) {
```
Comment (DevYukine) [inline commands/social/store/store.js]: you actuall do the same thing 2 times here, i would recommend creating 2 variable for each number.

**Review by DevYukine — CHANGES_REQUESTED** [review]
after this requested changes, this could be merged imo

**Review by DevYukine — APPROVED** [review]

**Pull request merged**